elixir/apps/domain/lib/domain/change_logs/replication_connection.ex

```diff
@@ -1,90 +1,87 @@
   end
 
-  defp log(_op, _lsn, "tokens", _old_data, %{"type" => "relay_group"}) do
-    :ok
+  def on_write(state, lsn, op, table, %{"account_id" => account_id} = old_data, data)
+      when not is_nil(account_id) do
+    buffer(state, lsn, op, table, account_id, old_data, data)
   end
 
-  defp log(op, lsn, table, old_data, data) do
-    attrs = %{
-      op: op,
+  # If we get here, raise the alarm as it means we encountered a change we didn't expect.
+  def on_write(state, lsn, op, table, _old_data, _data) do
+    Logger.error(
+      "Unexpected write operation!",
       lsn: lsn,
-      table: table,
-      old_data: old_data,
-      data: data,
-      vsn: @vsn
-    }
-
-    case ChangeLogs.create_change_log(attrs) do
-      {:ok, _change_log} ->
-        :ok
-
-      {:error, %Ecto.Changeset{errors: errors} = changeset} ->
-        if Enum.any?(errors, &should_skip_change_log?/1) do
-          # Expected under normal operation when an account is deleted or we are catching up on
-          # already-processed but not acknowledged WAL data.
-          :ok
-        else
-          Logger.warning("Failed to create change log",
-            errors: inspect(changeset.errors),
-            table: table,
-            op: op,
-            lsn: lsn,
-            vsn: @vsn
-          )
-
-          # TODO: WAL
-          # Don't ignore failures to insert change logs. Improve this after we have some
-          # operational experience with the data flowing in here.
-          :ok
-        end
-    end
-  end
+      op: op,
+      table: table
+    )
 
-  defp should_skip_change_log?({:account_id, {"does not exist", _violations}}) do
-    true
+    state
   end
 
-  defp should_skip_change_log?({:lsn, {"has already been taken", _violations}}) do
-    true
+  def on_flush(%{flush_buffer: flush_buffer} = state) when map_size(flush_buffer) == 0, do: state
+
+  def on_flush(state) do
+    to_insert = Map.values(state.flush_buffer)
+    attempted_count = Enum.count(state.flush_buffer)
+
+    {successful_count, _change_logs} = ChangeLogs.bulk_insert(to_insert)
+
+    Logger.info("Flushed #{successful_count}/#{attempted_count} change logs")
+
+    # We always advance the LSN to the highest LSN in the flush buffer because
+    # LSN conflicts just mean the data is already inserted, and other insert_all
+    # issues like a missing account_id will raise an exception.
+    last_lsn =
+      state.flush_buffer
+      |> Map.keys()
+      |> Enum.max()
+
+    %{state | flush_buffer: %{}, last_flushed_lsn: last_lsn}
   end
 
-  defp should_skip_change_log?(_error) do
-    false
+  defp buffer(state, lsn, op, table, account_id, old_data, data) do
+    flush_buffer =
+      state.flush_buffer
+      |> Map.put_new(lsn, %{
+        lsn: lsn,
+        op: op,
+        table: table,
+        account_id: account_id,
+        old_data: old_data,
+        data: data,
+        vsn: @vsn
+      })
+
+    %{state | flush_buffer: flush_buffer}
```

Comment:
Just realized, should we be using a GenServer for this?

Reply:
We are in effect - that's what the `use Domain.Replication.Connection` does.